JAVA-2611: DNS Lookup of SRV and TXT records

jyemin opened the pull request:
At the cost of making synchronous network calls in the ConnectionString constructor, this implementation of DNS lookup preserves backwards compatibility for any application that is currently using a connection string and then just dumps in a mongodb+src one in its place.

In the future we can implement an alternative API that doesn't have this constraint, but for now this is the safest way forward with the least surprises for application authors.

Patch build: https://evergreen.mongodb.com/version/59f250c8e3c3310b8f002b56

jyemin commented on driver-core/src/main/com/mongodb/internal/dns/DnsResolver.java:
```diff
@@ -0,0 +1,107 @@
+            while (srvRecordEnumeration.hasMore()) {
+                String srvRecord = (String) srvRecordEnumeration.next();
+                String[] split = srvRecord.split(" ");
+                String resolvedHost = split[3].endsWith(".") ? split[3].substring(0, split[3].length() - 1) : split[3];
```
Should it be an error if the resolve host does not end with a "."?

craiggwilson replied:
no... I think this is fine.

craiggwilson commented on driver-core/src/main/com/mongodb/internal/dns/DnsResolver.java:
```diff
@@ -0,0 +1,107 @@
+        return hosts;
+    }
+
+    // A TXT record is just a string
```
A TXT record can be multiple strings.

jyemin replied:
I mean that each one a single string, which are then combined into a single string with '&' separators.  Is that incorrect?

craiggwilson replied:
Yes, that is incorrect. Each TXT record can contain multiple strings.

jyemin replied:
The JNDI library combines the multiple strings for each TXT record (and adds a space between each).  So this loop is for combining multiple TXT records

rozza replied:
Ack

craiggwilson commented on driver-core/src/main/com/mongodb/ConnectionString.java:
```diff
@@ -793,6 +837,10 @@ private void validatePort(final String host, final String port) {
             throw new IllegalArgumentException(format("The connection string contains an invalid host '%s'. "
                     + "The port '%s' is not a valid, it must be an integer between 0 and 65535", host, port));
         }
+        if (isSRVProtocol) {
```
any reason not to check this at the beginning before parsing?

jyemin commented:
Added a couple more commits:

* Updated to latest specification tests, including those in the PR for SPEC-984
* Changed implementation to abide by SPEC-984 (readPreferenceTags issue)

jyemin commented:
Note: we're awaiting a build ticket to add the new SRV/TXT records to DNS, so until then three of the spec tests are failing.  I'll keep this open until I can verify that all the tests pass.

jyemin commented on driver-core/src/main/com/mongodb/ConnectionString.java:
```diff
@@ -349,9 +349,10 @@ public ConnectionString(final String connectionString) {
-        warnOnUnsupportedOptions(connectionStringOptionsMap);
+        Map<String, List<String>> combinedOptionsMaps = combineOptionsMaps(txtRecordsOptionsMap, connectionStringOptionsMap);
+        translateOptions(combinedOptionsMaps);
+        credential = createCredentials(combinedOptionsMaps, userName, password);
```
No way to test this change currently because none of the TXT records have auth information in them (e.g a authSource)

jyemin commented on driver-core/src/main/com/mongodb/internal/dns/DnsResolver.java:
```diff
@@ -76,7 +76,10 @@ public static String resolveAdditionalQueryParametersFromTxtRecords(final String
                 NamingEnumeration<?> txtRecordEnumeration = attribute.getAll();
                 while (txtRecordEnumeration.hasMore()) {
-                    String txtRecord = (String) txtRecordEnumeration.next();
+                    // Remove all space characters, as the DNS resolver for TXT records inserts a space character
```
I think this is legit, as URIs shouldn't have their own spaces in them.

jyemin commented on driver-core/src/test/resources/initial-dns-seedlist-discovery/txt-record-unknown-key.json:
```diff
@@ -1,7 +0,0 @@
-{
-  "uri": "mongodb+srv://test9.test.build.10gen.cc/?replicaSet=repl0",
```
SPEC-995 removes this

jyemin commented on driver-core/src/test/resources/initial-dns-seedlist-discovery/txt-record-utf8-character.json:
```diff
@@ -1,20 +0,0 @@
-{
```
This one is likely to be removed as well.  Can add it back if we work out how to do it.

jyemin commented:
New patch
